File: python/python2_7_18/Java/PythonLexerBase.java
/*
The MIT License (MIT)
Copyright (c) 2021 Robert Einhorn

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:
The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
 */

/*
 *
 * Project      : Python Indent/Dedent handler for ANTLR4 grammars
 *
 * Developed by : Robert Einhorn, robert.einhorn.hu@gmail.com
 *
 */

 import java.util.*;

 import org.antlr.v4.runtime.*;
 
 public abstract class PythonLexerBase extends Lexer {
     // A stack that keeps track of the indentation lengths
     private Deque<Integer> indentLengthStack;
     // A list where tokens are waiting to be loaded into the token stream
     private LinkedList<Token> pendingTokens;
 
     // last pending token types
     private int previousPendingTokenType;
     private int lastPendingTokenTypeFromDefaultChannel;
 
     // The amount of opened parentheses, square brackets or curly braces
     private int opened;
      
     private boolean wasSpaceIndentation;
     private boolean wasTabIndentation;
     private boolean wasIndentationMixedWithSpacesAndTabs;
 
     private Token curToken; // current (under processing) token
     private Token ffgToken; // following (look ahead) token
 
     private final int INVALID_LENGTH = -1;
     private final String ERR_TXT = " ERROR: ";
 
     protected PythonLexerBase(CharStream input) {
         super(input);
         this.init();
     }
 
     @Override
     public Token nextToken() { // reading the input stream until a return EOF
         this.checkNextToken();
         return this.pendingTokens.pollFirst(); // add the queued token to the token stream
     }
 
     @Override
     public void reset() {
         this.init();
         super.reset();
     }
 
     private void init() {
         this.indentLengthStack = new ArrayDeque<>();
         this.pendingTokens = new LinkedList<>();
         this.previousPendingTokenType = 0;
         this.lastPendingTokenTypeFromDefaultChannel = 0;
         this.opened = 0;
         this.wasSpaceIndentation = false;
         this.wasTabIndentation = false;
         this.wasIndentationMixedWithSpacesAndTabs = false;
         this.curToken = null;
         this.ffgToken = null;
     }
 
     private void checkNextToken() {
         if (this.previousPendingTokenType != Token.EOF) {
             this.setCurrentAndFollowingTokens();
             if (this.indentLengthStack.isEmpty()) { // We're at the first token
                 this.handleStartOfInput();
             }
 
             switch (this.curToken.getType()) {
                 case PythonLexer.LPAR:
                 case PythonLexer.LSQB:
                 case PythonLexer.LBRACE:
                     this.opened++;
                     this.addPendingToken(this.curToken);
                     break;
                 case PythonLexer.RPAR:
                 case PythonLexer.RSQB:
                 case PythonLexer.RBRACE:
                     this.opened--;
                     this.addPendingToken(this.curToken);
                     break;
                 case PythonLexer.NEWLINE:
                     this.handleNEWLINEtoken();
                     break;
                 case PythonLexer.ERRORTOKEN:
                     this.reportLexerError("token recognition error at: '" + this.curToken.getText() + "'");
                     this.addPendingToken(this.curToken);
                     break;
                 case Token.EOF:
                     this.handleEOFtoken();
                     break;
                 default:
                     this.addPendingToken(this.curToken);
             }
         }
     }
 
     private void setCurrentAndFollowingTokens() {
         this.curToken = this.ffgToken == null ?
                         super.nextToken() :
                         this.ffgToken;
 
          this.ffgToken = this.curToken.getType() == Token.EOF ?
                         this.curToken :
                         super.nextToken();
     }
 
     // initialize the indentLengthStack
     // hide the leading NEWLINE token(s)
     // if exists, find the first statement (not NEWLINE, not EOF token) that comes from the default channel
     // insert a leading INDENT token if necessary
     private void handleStartOfInput() {
         // initialize the stack with a default 0 indentation length
         this.indentLengthStack.push(0); // this will never be popped off
         while (this.curToken.getType() != Token.EOF) {
             if (this.curToken.getChannel() == Token.DEFAULT_CHANNEL) {
                 if (this.curToken.getType() == PythonLexer.NEWLINE) {
                     // all the NEWLINE tokens must be ignored before the first statement
                     this.hideAndAddPendingToken(this.curToken);
                 } else { // We're at the first statement
                     this.insertLeadingIndentToken();
                     return; // continue the processing of the current token with checkNextToken()
                 }
             } else {
                 this.addPendingToken(this.curToken); // it can be WS, EXPLICIT_LINE_JOINING or COMMENT token
             }
             this.setCurrentAndFollowingTokens();
         }
         // continue the processing of the EOF token with checkNextToken()
     }
 
     private void insertLeadingIndentToken() {
         if (this.previousPendingTokenType == PythonLexer.WS) {
             Token prevToken = this.pendingTokens.peekLast(); // WS token
             if (this.getIndentationLength(prevToken.getText()) != 0) { // there is an "indentation" before the first statement
                 final String errMsg = "first statement indented";
                 this.reportLexerError(errMsg);
                 // insert an INDENT token before the first statement to raise an 'unexpected indent' error later by the parser
                 this.createAndAddPendingToken(PythonLexer.INDENT, Token.DEFAULT_CHANNEL, this.ERR_TXT + errMsg, this.curToken);
             }
         }
     }
 
     private void handleNEWLINEtoken() {
         if (this.opened > 0) { // We're in an implicit line joining, ignore the current NEWLINE token
             this.hideAndAddPendingToken(this.curToken);
         } else {
             final Token nlToken = new CommonToken(this.curToken); // save the current NEWLINE token
             final boolean isLookingAhead = this.ffgToken.getType() == PythonLexer.WS;
             if (isLookingAhead) {
                 this.setCurrentAndFollowingTokens(); // set the next two tokens
             }
 
             switch (this.ffgToken.getType()) {
                 case PythonLexer.NEWLINE: // We're before a blank line
                 case PythonLexer.COMMENT: // We're before a comment
                     this.hideAndAddPendingToken(nlToken);
                     if (isLookingAhead) {
                         this.addPendingToken(this.curToken); // WS token
                     }
                     break;
                 default:
                     this.addPendingToken(nlToken);
                     if (isLookingAhead) { // We're on whitespace(s) followed by a statement
                         final int indentationLength = this.ffgToken.getType() == Token.EOF ?
                                                       0 :
                                                       this.getIndentationLength(this.curToken.getText());
 
                         if (indentationLength != this.INVALID_LENGTH) {
                             this.addPendingToken(this.curToken); // WS token
                             this.insertIndentOrDedentToken(indentationLength); // may insert INDENT token or DEDENT token(s)
                         } else {
                             this.reportError("inconsistent use of tabs and spaces in indentation");
                         }
                     } else { // We're at a newline followed by a statement (there is no whitespace before the statement)
                         this.insertIndentOrDedentToken(0); // may insert DEDENT token(s)
                     }
             }
         }
     }
 
     private void insertIndentOrDedentToken(final int indentLength) {
         int prevIndentLength = this.indentLengthStack.peek();
         if (indentLength > prevIndentLength) {
             this.createAndAddPendingToken(PythonLexer.INDENT, Token.DEFAULT_CHANNEL, null, this.ffgToken);
             this.indentLengthStack.push(indentLength);
         } else {
             while (indentLength < prevIndentLength) { // more than 1 DEDENT token may be inserted to the token stream
                 this.indentLengthStack.pop();
                 prevIndentLength = this.indentLengthStack.peek();
                 if (indentLength <= prevIndentLength) {
                     this.createAndAddPendingToken(PythonLexer.DEDENT, Token.DEFAULT_CHANNEL, null, this.ffgToken);
                 } else {
                     this.reportError("inconsistent dedent");
                 }
             }
         }
     }
 
     private void insertTrailingTokens() {
         switch (this.lastPendingTokenTypeFromDefaultChannel) {
             case PythonLexer.NEWLINE:
             case PythonLexer.DEDENT:
                 break; // no trailing NEWLINE token is needed
             default:
                 // insert an extra trailing NEWLINE token that serves as the end of the last statement
                 this.createAndAddPendingToken(PythonLexer.NEWLINE, Token.DEFAULT_CHANNEL, null, this.ffgToken); // ffgToken is EOF
         }
         this.insertIndentOrDedentToken(0); // Now insert as much trailing DEDENT tokens as needed
     }
 
     private void handleEOFtoken() {
         if (this.lastPendingTokenTypeFromDefaultChannel > 0) {
             // there was statement in the input (leading NEWLINE tokens are hidden)
             this.insertTrailingTokens();
         }
         this.addPendingToken(this.curToken);
     }
 
     private void hideAndAddPendingToken(final Token tkn) {
         CommonToken ctkn = new CommonToken(tkn);
         ctkn.setChannel(Token.HIDDEN_CHANNEL);
         this.addPendingToken(ctkn);
     }
 
     private void createAndAddPendingToken(final int ttype, final int channel, final String text, Token sampleToken) {
         CommonToken ctkn = new CommonToken(sampleToken);
         ctkn.setType(ttype);
         ctkn.setChannel(channel);
         ctkn.setStopIndex(sampleToken.getStartIndex() - 1);
         ctkn.setText(text == null
                      ? "<" + this.getVocabulary().getDisplayName(ttype) + ">"
                      : text);
 
         this.addPendingToken(ctkn);
     }
 
     private void addPendingToken(final Token tkn) {
         // save the last pending token type because the pendingTokens linked list can be empty by the nextToken()
         this.previousPendingTokenType = tkn.getType();
         if (tkn.getChannel() == Token.DEFAULT_CHANNEL) {
             this.lastPendingTokenTypeFromDefaultChannel = this.previousPendingTokenType;
         }
         this.pendingTokens.addLast(tkn);
     }
 
     private int getIndentationLength(final String indentText) { // the indentText may contain spaces, tabs or form feeds
         final int TAB_LENGTH = 8; // the standard number of spaces to replace a tab to spaces
         int length = 0;
         for (char ch : indentText.toCharArray()) {
             switch (ch) {
                 case ' ':
                     this.wasSpaceIndentation = true;
                     length += 1;
                     break;
                 case '\t':
                     this.wasTabIndentation = true;
                     length += TAB_LENGTH - (length % TAB_LENGTH);
                     break;
                 case '\f': // form feed
                     length = 0;
                     break;
             }
         }
 
         if (this.wasTabIndentation && this.wasSpaceIndentation) {
             if (!(this.wasIndentationMixedWithSpacesAndTabs)) {
                 this.wasIndentationMixedWithSpacesAndTabs = true;
                 length = this.INVALID_LENGTH; // only for the first inconsistent indent
             }
         }
         return length;
     }
 
     private void reportLexerError(final String errMsg) {
         this.getErrorListenerDispatch().syntaxError(this, this.curToken, this.curToken.getLine(), this.curToken.getCharPositionInLine(), " LEXER" + this.ERR_TXT + errMsg, null);
     }
 
     private void reportError(final String errMsg) {
         this.reportLexerError(errMsg);
 
         // the ERRORTOKEN will raise an error in the parser
         this.createAndAddPendingToken(PythonLexer.ERRORTOKEN, Token.DEFAULT_CHANNEL, this.ERR_TXT + errMsg, this.ffgToken);
     }
 }
 

File: python/python2_7_18/Python3/PythonLexerBase.py
# The MIT License (MIT)
# Copyright (c) 2021 Robert Einhorn
# 
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.

# Project      : Python Indent/Dedent handler for ANTLR4 grammars
# 
# Developed by : Robert Einhorn

from typing import TextIO
from antlr4 import InputStream, Lexer, Token
from antlr4.Token import CommonToken
import sys
import re

class PythonLexerBase(Lexer):
    def __init__(self, input: InputStream, output: TextIO = sys.stdout):
        super().__init__(input, output)

        # A stack that keeps track of the indentation lengths
        self.__indent_length_stack: list[int]

        # A list where tokens are waiting to be loaded into the token stream
        self.__pending_tokens: list[CommonToken]

        # last pending token types
        self.__previous_pending_token_type: int
        self.__last_pending_token_type_from_default_channel: int

        # The amount of opened parentheses, square brackets or curly braces
        self.__opened: int

        self.__was_space_indentation: bool
        self.__was_tab_indentation: bool
        self.__was_indentation_mixed_with_spaces_and_tabs: bool

        self.__cur_token: CommonToken # current (under processing) token
        self.__ffg_token: CommonToken # following (look ahead) token

        self.__INVALID_LENGTH: int = -1
        self.__ERR_TXT: str = " ERROR: "

        self.__init()

    def nextToken(self) -> CommonToken: # reading the input stream until a return EOF
        self.__check_next_token()
        return self.__pending_tokens.pop(0) # add the queued token to the token stream

    def reset(self) -> None:
        self.__init()
        super().reset()

    def __init(self) -> None:
        self.__indent_length_stack = []
        self.__pending_tokens = []
        self.__previous_pending_token_type = 0
        self.__last_pending_token_type_from_default_channel = 0
        self.__opened = 0
        self.__was_space_indentation = False
        self.__was_tab_indentation = False
        self.__was_indentation_mixed_with_spaces_and_tabs = False
        self.__cur_token = None
        self.__ffg_token = None

    def __check_next_token(self) -> None:
        if self.__previous_pending_token_type != Token.EOF:
            self.__set_current_and_following_tokens()
            if len(self.__indent_length_stack) == 0: # We're at the first token
                self.__handle_start_of_input()

            match self.__cur_token.type:
                case self.LPAR | self.LSQB | self.LBRACE:
                    self.__opened += 1
                    self.__add_pending_token(self.__cur_token)
                case self.RPAR | self.RSQB | self.RBRACE:
                    self.__opened -= 1
                    self.__add_pending_token(self.__cur_token)
                case self.NEWLINE:
                    self.__handle_NEWLINE_token()
                case self.ERRORTOKEN:
                    self.__report_lexer_error("token recognition error at: '" + self.__cur_token.text + "'")
                    self.__add_pending_token(self.__cur_token)
                case Token.EOF:
                    self.__handle_EOF_token()
                case other:
                    self.__add_pending_token(self.__cur_token)

    def __set_current_and_following_tokens(self) -> None:
        self.__cur_token = super().nextToken() if self.__ffg_token is None else \
                            self.__ffg_token

        self.__ffg_token = self.__cur_token if self.__cur_token.type == Token.EOF else \
                            super().nextToken()

    # initialize the _indent_length_stack
    # hide the leading NEWLINE token(s)
    # if exists, find the first statement (not NEWLINE, not EOF token) that comes from the default channel
    # insert a leading INDENT token if necessary
    def __handle_start_of_input(self) -> None:
        # initialize the stack with a default 0 indentation length
        self.__indent_length_stack.append(0) # this will never be popped off
        while self.__cur_token.type != Token.EOF:
            if self.__cur_token.channel == Token.DEFAULT_CHANNEL:
                if self.__cur_token.type == self.NEWLINE:
                    # all the NEWLINE tokens must be ignored before the first statement
                    self.__hide_and_add_pending_token(self.__cur_token)
                else: # We're at the first statement
                    self.__insert_leading_indent_token()
                    return # continue the processing of the current token with __check_next_token()
            else:
                self.__add_pending_token(self.__cur_token) # it can be WS, EXPLICIT_LINE_JOINING or COMMENT token
            self.__set_current_and_following_tokens()
        # continue the processing of the EOF token with __check_next_token()

    def __insert_leading_indent_token(self) -> None:
        if self.__previous_pending_token_type == self.WS:
            prev_token: CommonToken = self.__pending_tokens[-1]  # WS token
            if self.__get_indentation_length(prev_token.text) != 0: # there is an "indentation" before the first statement
                err_msg: str = "first statement indented"
                self.__report_lexer_error(err_msg)
                # insert an INDENT token before the first statement to raise an 'unexpected indent' error later by the parser
                self.__create_and_add_pending_token(self.INDENT, Token.DEFAULT_CHANNEL, self.__ERR_TXT + err_msg, self.__cur_token)

    def __handle_NEWLINE_token(self) -> None:
        if self.__opened > 0: # We're in an implicit line joining, ignore the current NEWLINE token
            self.__hide_and_add_pending_token(self.__cur_token)
        else:
            nl_token: CommonToken = self.__cur_token.clone() # save the current NEWLINE token
            is_looking_ahead: bool = self.__ffg_token.type == self.WS
            if is_looking_ahead:
                self.__set_current_and_following_tokens() # set the next two tokens

            match self.__ffg_token.type:
                case self.NEWLINE | self.COMMENT:
                    # We're before a blank line or a comment or type comment or a type ignore comment
                    self.__hide_and_add_pending_token(nl_token) # ignore the NEWLINE token
                    if is_looking_ahead:
                        self.__add_pending_token(self.__cur_token) # WS token
                case other:
                    self.__add_pending_token(nl_token)
                    if is_looking_ahead: # We're on a whitespace(s) followed by a statement
                        indentation_length: int = 0 if self.__ffg_token.type == Token.EOF else \
                                                  self.__get_indentation_length(self.__cur_token.text)

                        if indentation_length != self.__INVALID_LENGTH:
                            self.__add_pending_token(self.__cur_token) # WS token
                            self.__insert_indent_or_dedent_token(indentation_length) # may insert INDENT token or DEDENT token(s)
                        else:
                            self.__report_error("inconsistent use of tabs and spaces in indentation")
                    else: # We're at a newline followed by a statement (there is no whitespace before the statement)
                        self.__insert_indent_or_dedent_token(0) # may insert DEDENT token(s)

    def __insert_indent_or_dedent_token(self, indent_length: int) -> None:
        prev_indent_length: int = self.__indent_length_stack[-1] # peek()
        if indent_length > prev_indent_length:
            self.__create_and_add_pending_token(self.INDENT, Token.DEFAULT_CHANNEL, None, self.__ffg_token)
            self.__indent_length_stack.append(indent_length)
        else:
            while indent_length < prev_indent_length: # more than 1 DEDENT token may be inserted to the token stream
                self.__indent_length_stack.pop()
                prev_indent_length = self.__indent_length_stack[-1] # peek()
                if indent_length <= prev_indent_length:
                    self.__create_and_add_pending_token(self.DEDENT, Token.DEFAULT_CHANNEL, None, self.__ffg_token)
                else:
                    self.__report_error("inconsistent dedent")

    def __insert_trailing_tokens(self) -> None:
        match self.__last_pending_token_type_from_default_channel:
            case self.NEWLINE | self.DEDENT:
                pass # no trailing NEWLINE token is needed
            case other:
                # insert an extra trailing NEWLINE token that serves as the end of the last statement
                self.__create_and_add_pending_token(self.NEWLINE, Token.DEFAULT_CHANNEL, None, self.__ffg_token) # _ffg_token is EOF
        self.__insert_indent_or_dedent_token(0) # Now insert as much trailing DEDENT tokens as needed

    def __handle_EOF_token(self) -> None:
        if self.__last_pending_token_type_from_default_channel > 0:
            # there was statement in the input (leading NEWLINE tokens are hidden)
            self.__insert_trailing_tokens()
        self.__add_pending_token(self.__cur_token)

    def __hide_and_add_pending_token(self, ctkn: CommonToken) -> None:
        ctkn.channel = Token.HIDDEN_CHANNEL
        self.__add_pending_token(ctkn)

    def __create_and_add_pending_token(self, ttype: int, channel: int, text: str, sample_token: CommonToken) -> None:
        ctkn: CommonToken = sample_token.clone()
        ctkn.type  = ttype
        ctkn.channel = channel
        ctkn.stop = sample_token.start - 1
        ctkn.text = "<" + self.symbolicNames[ttype] + ">" if text is None else \
                        text

        self.__add_pending_token(ctkn)

    def __add_pending_token(self, ctkn: CommonToken) -> None:
        # save the last pending token type because the _pending_tokens list can be empty by the nextToken()
        self.__previous_pending_token_type = ctkn.type
        if ctkn.channel == Token.DEFAULT_CHANNEL:
            self.__last_pending_token_type_from_default_channel = self.__previous_pending_token_type
        self.__pending_tokens.append(ctkn)

    def __get_indentation_length(self, indentText: str) -> int: # the indentText may contain spaces, tabs or form feeds
        TAB_LENGTH: int = 8 # the standard number of spaces to replace a tab to spaces
        length: int = 0
        ch: str
        for ch in indentText:
            match ch:
                case ' ':
                    self.__was_space_indentation = True
                    length += 1
                case '\t':
                    self.__was_tab_indentation = True
                    length += TAB_LENGTH - (length % TAB_LENGTH)
                case '\f': # form feed
                    length = 0

        if self.__was_tab_indentation and self.__was_space_indentation:
            if not self.__was_indentation_mixed_with_spaces_and_tabs:
                self.__was_indentation_mixed_with_spaces_and_tabs = True
                length = self.__INVALID_LENGTH # only for the first inconsistent indent
        return length

    def __report_lexer_error(self, err_msg: str) -> None:
        self.getErrorListenerDispatch().syntaxError(self, self.__cur_token, self.__cur_token.line, self.__cur_token.column, " LEXER" + self.__ERR_TXT + err_msg, None)

    def __report_error(self, err_msg: str) -> None:
        self.__report_lexer_error(err_msg)

        # the ERRORTOKEN will raise an error in the parser
        self.__create_and_add_pending_token(self.ERRORTOKEN, Token.DEFAULT_CHANNEL, self.__ERR_TXT + err_msg, self.__ffg_token)
